crates/biome_css_parser/src/syntax/at_rule/keyframes.rs

```diff
@@ -103,11 +114,34 @@ impl ParseNodeList for KeyframesItemList {
         p: &mut Self::Parser<'_>,
         parsed_element: ParsedSyntax,
     ) -> RecoveryResult {
-        parsed_element.or_recover_with_token_set(
-            p,
-            &ParseRecoveryTokenSet::new(CSS_BOGUS_KEYFRAMES_ITEM, KEYFRAMES_ITEM_LIST_RECOVERY_SET),
-            expected_keyframes_item,
-        )
+        parsed_element.or_recover(p, &KeyframesItemListParseRecovery, expected_keyframes_item)
+    }
+}
+
+struct KeyframesItemBlockParseRecovery;
+
+impl ParseRecovery for KeyframesItemBlockParseRecovery {
+    type Kind = CssSyntaxKind;
+    type Parser<'source> = CssParser<'source>;
+    const RECOVERED_KIND: Self::Kind = CSS_BOGUS_BLOCK;
+
+    fn is_at_recovered(&self, p: &mut Self::Parser<'_>) -> bool {
+        // We need to recover the contents of an invalid block for the following cases:
+        // We have a declaration block, but it lacks a '{' token:
+        //    @keyframes name {
+        //      from
+        //          color: red;
+        //      } <----- here it's a recover point for the keyframes item block.
+        //   }
+        //
+        //    @keyframes name {
+        //      from
+        //          color: red;
+        //      to <----- here it's a recover point for the next keyframes item. {
+        //          color: blue;
+        //      }
+        //   }
+        p.at_ts(token_set!(T!['}'])) || is_at_keyframes_item_selector(p)
     }
```

Comment:
very nice! i love how easy this looks with the new parse recovery setup.

Reply:
It's very difficult to determine how to create the correct recovery parsing. =( 
It's becoming clearer, however, I'm still unsure about the right way to handle it.